### PR TITLE
Add integration/IR test cases for gather

### PR DESCRIPTION
**Pull request by Mgluhovskoi**
Addresses Issue #52.

### CODE REVIEW

File: tripy/tests/integration/test_gather.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+import tripy as tp
+
+
+class TestGatherOp:
+    @pytest.mark.parametrize(
+        "x_shape, axis, indices",
+        [
+            ((2, 3), 0, (1)),
+            pytest.param((2, 3, 4), 1, (1, 2),marks=pytest.mark.xfail(reason="Bug in gather op. (Remove xfail when fixed)")),
```
Comment (parthchadha) [inline tripy/tests/integration/test_gather.py:30]: What's the bug? Can you file an issue with details? (1 issue for all failures)

Comment (Mgluhovskoi) [reply tripy/tests/integration/test_gather.py:30]: I filed an issue on gitlab should I file one on github instead?

Comment (parthchadha) [reply tripy/tests/integration/test_gather.py:30]: Github.

File: tripy/tests/flat_ir/ops/test_gather.py [hunk 2]
```diff
@@ -16,28 +16,37 @@
...
+import pytest
 
 class TestGatherOp:
-    def test_gather_str(self):
-        data = tp.Tensor([3.0, 4.0], name="data")
-        index = tp.Tensor([0], dtype=int32, name="indices")
-        out = tp.gather(data, 0, index)
+    @pytest.mark.parametrize("axis", [0,1,2])
+    def test_gather_str_dim0(self, axis):
+        data = tp.Tensor([[[0.0000, 0.0000],
```
Comment (pranavm-nvidia) [inline tripy/tests/flat_ir/ops/test_gather.py:27]: Seems like this can be generated via `iota`

File: tripy/tests/flat_ir/ops/test_gather.py [hunk 3]
```diff
@@ -16,28 +16,37 @@
...
 
         gather = flat_ir.ops[-1]
         reshape = flat_ir.ops[-2]
-
         print(str(reshape))
         assert isinstance(gather, DynamicGatherOp)
         assert (
             str(gather)
-            == "out: [rank=(1), dtype=(float32), loc=(gpu:0)] = DynamicGatherOp(data, indices, t_inter3, axis=0)"
+            == f"out: [rank=(3), dtype=(float32), loc=(gpu:0)] = DynamicGatherOp(data, indices, t_inter3, axis={axis})"
```
Comment (pranavm-nvidia) [inline tripy/tests/flat_ir/ops/test_gather.py:51]: I think the more important thing to test is the FlatIR->MLIR translation. Can you add a test for that? You can look at some of the other files in this directory for examples.

Comment (pranavm-nvidia) [reply tripy/tests/flat_ir/ops/test_gather.py]: See `tripy/tests/flat_ir/ops/test_iota.py`

Comment (Mgluhovskoi) [reply tripy/tests/flat_ir/ops/test_gather.py]: I added some test cases, but I am not sure how much of the MLIR should be checked.

Comment (pranavm-nvidia) [reply tripy/tests/flat_ir/ops/test_gather.py]: Oh my mistake, looks like the fix for non-zero axis was actually in the trace op. Could you add a test to check the trace->FlatIR translation? We should make sure we're generating the right set of ops. @parthchadha could probably provide more details on what exactly we should look out for

Comment (pranavm-nvidia) [reply tripy/tests/flat_ir/ops/test_gather.py]: Oh my mistake, looks like the fix for non-zero axis was actually in the trace op. Could you add a test to check the trace->FlatIR translation? We should make sure we're generating the right set of ops. @parthchadha could probably provide more details on what exactly we should look out for. The PR description says it fixes the case where `axis > 0`, but the code seems like it's for the case where `axis == rank - 1`. Not entirely sure I've understood what's happening there.

Comment (Mgluhovskoi) [reply tripy/tests/flat_ir/ops/test_gather.py]: Which PR are you referring to? And I thought the test case that I wrote already checks the translation from trace->FlatIR.

Comment (pranavm-nvidia) [reply tripy/tests/flat_ir/ops/test_gather.py]: This commit: https://github.com/NVIDIA/TensorRT-Incubator/commit/c0784f34844f65600f95038555f44c2b52b1cfd1
It does implicitly test trace->FlatIR, but I think we should explicitly test it also in `tripy/tests/frontend/trace/ops/test_gather.py`

Comment (parthchadha) [reply tripy/tests/flat_ir/ops/test_gather.py]: The old was completely broken when axis > 0 (slicing on the shape tensor was incorrect for `slice_sizes`)

File: tripy/tests/flat_ir/ops/test_gather.py [hunk 4]
```diff
@@ -16,28 +16,41 @@
 #
 
 import tripy as tp
-from tripy import int32
 
 from tripy.flat_ir.ops import DynamicGatherOp
 from tripy.frontend.trace import Trace
-
+import pytest
```
Comment (jhalakpatel) [inline tripy/tests/flat_ir/ops/test_gather.py:22]: This should go above `import tripy as tp`.

File: tripy/tests/flat_ir/ops/test_gather.py [hunk 5]
```diff
@@ -16,28 +16,41 @@
...
         assert isinstance(gather, DynamicGatherOp)
         assert (
             str(gather)
-            == "out: [rank=(1), dtype=(float32), loc=(gpu:0)] = DynamicGatherOp(data, indices, t_inter3, axis=0)"
+            == f"out: [rank=(3), dtype=(float32), loc=(gpu:0)] = DynamicGatherOp(data, indices, t_inter4, axis={axis})"
         )
+
+    @pytest.mark.parametrize("axis", [0,1])
+    def test_gather_mlir(self, axis):
+        out = tp.gather(tp.Tensor([[1,1,1],[1,1,1]]),axis,tp.Tensor([0],dtype=tp.int32))
```
Comment (jhalakpatel) [inline tripy/tests/flat_ir/ops/test_gather.py:47]: Looks like the pre-commit hook for formatting was not installed?

File: tripy/tests/flat_ir/ops/test_gather.py [hunk 6]
```diff
@@ -16,28 +16,41 @@
...
-
+import pytest
 
 class TestGatherOp:
-    def test_gather_str(self):
-        data = tp.Tensor([3.0, 4.0], name="data")
-        index = tp.Tensor([0], dtype=int32, name="indices")
-        out = tp.gather(data, 0, index)
+    @pytest.mark.parametrize("axis", [0,1,2])
+    def test_gather_str_dim0(self, axis):
```
Comment (jhalakpatel) [inline tripy/tests/flat_ir/ops/test_gather.py:26]: What does the suffix `dim0` signify?

Comment (Mgluhovskoi) [reply tripy/tests/flat_ir/ops/test_gather.py:26]: Was planning on doing separate test cases for each dimension tested. I will remove it.

**Comment by pranavm-nvidia** [comment]
@Mgluhovskoi seems like your pre-commit hook's line length is set to the default value of 80 instead of 120 like we usually use?

File: tripy/tests/flat_ir/ops/test_gather.py (hunk 3 above)
Comment (Mgluhovskoi) [reply tripy/tests/flat_ir/ops/test_gather.py]: Added more trace->FlatIR testing for when axis>0.

**Review by parthchadha — APPROVED** [review]

**Pull request merged**